### ISSUE TITLE
Computed properties

### ISSUE THREAD

**Issue by nolanlawson** [salesforce/lwc#2437]
Related: #243 and https://github.com/salesforce/lwc/issues/2266#issuecomment-792838789

## Justification

In [Vue](https://v3.vuejs.org/guide/computed.html#computed-properties) and [Svelte](https://v2.svelte.dev/guide#computed-properties) there is a concept of **computed properties**, where a property can be derived from other properties.

[Example in Vue](https://codepen.io/nolanlawson-the-selector/pen/eYWVrJW):

```vue
<template>
  <span>{{a}} + {{b}} = {{sum}}</span>
</template>
<script>
export default {
  data() {
    return { a: 1, b: 2 }
  },
  computed: {
    sum () {
      return this.a + this.b
    }
  }
}
</script>
```

[Example in Svelte](https://svelte.dev/repl/aa75f620c7864f99b07fd1e8f6e5e79b?version=3.41.0):

```svelte
<span>{a} + {b} = {sum}</span>
<script>
  let a = 1
  let b = 2
  let sum
  $: sum = a + b
</script>
```

## Benefits

The main benefits of computed properties are:

1. **Avoiding boilerplate**: writing code to update the `sum` when `a` or `b` changes is unnecessary.
2. **Code clarity**: the relationship between `sum`, `a`, and `b` is clear.
3. **Performance**: the sum is only recomputed once when either `a` or `b` changes. (For summing numbers, this is trivial, but you can imagine more complicated computations.)

## Current solutions

Today, LWC developers would have to express computed properties using a pattern [like the following](https://webcomponents.dev/edit/484GoXRxGvEsRhyQYEEt):

```js
export default class extends LightningElement {
  _a = 1
  _b = 2
  sum = this._a + this._b

  get a () {
    return this._a
  }

  get b () {
    return this._b
  }

  set a (val) {
    this._a = val
    this.sum = this._a + this._b // recompute
  }

  set b (val) {
    this._b = val
    this.sum = this._a + this._b // recompute
  }  
}
```

As you can see, this is a lot of boilerplate. And plus, we are computing the sum twice if `a` and `b` both change. For computations that have many inputs, this can lead to many unnecessary computations. (Of course the developer can work around that, but it's even more boilerplate.)

https://github.com/salesforce/lwc/issues/243 would memoize the getters when reused in the same template, but it wouldn't handle every usage of a component getter. For instance, if the getter is called within a `renderedCallback`, it would not handle that.

## API sketch

Clearly we would need an RFC for this, but here is a basic sketch of what the API could look like:

```js
export default class extends LightningElement {
  a = 1
  b = 2

  @computed
  get sum () {
    return this.a + this.b
  }
}
```

**Comment by caridy** [comment]
Well, there are two things compounded here:

1. how to represent a computed value in the UI (which works without extra decoration in LWC). In the example above, if you want `sum` to show up in the UI, you have to do nothing, because the template reacts to a or b changing.
2. how to recompute a value that needs to used programatically. E.g.: using `sum` in the `onclick` even handlers. I don't think `sum` in the renderedCallback qualifies, unless you're not using `sum` in the template though.

Since 1 just work, the focus should be on 2 explicitly. We do have the internal mechanism for such reactivity mechanism, but I do feel that this can be solved more on the plugin system that we talked the other day, specifically, a reactive mechanism that is plugable, and can be used programmatically (module body) or declarative (template), I'm eager to explore that. Something like the following seems doable with the same plugin system that I mentioned the other day:

```
import { compute } from 'myLwc/computePlugin';
export default class extends LightningElement {
  a = 1
  b = 2
  computed;

  constructor() {
     super();
     this.computed = compute({
        sum () {
           return this.a + this.b
        }
    });
  }
}
```

Where `compute` can be a user-land plugin that taps into the reactive mechanism of a config object passed around (the argument) in the context of the component that is being created (a la hooks in react).

**Comment by nolanlawson** [comment]
@caridy You're right about # 1; I didn't think it through very well. A developer can do:

```js
export default class extends LightningElement {
  a = 1
  b = 2

  get sum() {
    return this.a + this.b
  }
}
```

And this [works fine](https://webcomponents.dev/edit/XBEOUNHfgCHAQSRuQJiG/) today – it responds to updates to `a` and `b`, and I even confirmed that it only does the computation once when both `a` and `b` change.

So the pattern I mentioned (the one with the setters) only really applies when the developer wants to avoid truly excessive re-computations, e.g. where the getter is called multiple times in an `onclick` or `renderedCallback` or anywhere else.

I like the idea of using the plugin system for this.
